pkg/yum/yum.go

```diff
@@ -44,32 +56,157 @@ type Location struct {
+	RequestTimeout time.Duration // Maximum time to wait for response while fetching
+}
+
+// NewRepository returns a new Repository whose URL is url
+func NewRepository(client http.Client, url string, timeout time.Duration) Repository {
+	return Repository{
+		URL:            url,
+		Packages:       nil,
+		Repomd:         Repomd{},
+		Client:         client,
+		RequestTimeout: timeout,
+	}
+}
+
+// FetchRepomd populates r.Repomd with repository's repomd.xml metadata. Returns response code and error.
+func (r *Repository) FetchRepomd() (int, error) {
+	var result Repomd
 	var err error
+	var resp *http.Response
+	var repomdURL string
 
-	if repomd, err = GetRepomdXML(client, url); err != nil {
-		return []Package{}, fmt.Errorf("Error parsing repomd.xml: %v", err)
+	if repomdURL, err = r.getRepomdURL(); err != nil {
+		return 0, fmt.Errorf("GET error: %w", err)
+	}
+	if resp, err = fetchWithTimeout(repomdURL, r.RequestTimeout); err != nil {
+		return 0, fmt.Errorf("GET error: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return resp.StatusCode, fmt.Errorf("Status error: %v", resp.StatusCode)
+	}
+	if result, err = parseRepomdXML(resp.Body); err != nil {
+		resp.Body.Close()
+		return resp.StatusCode, fmt.Errorf("error parsing repomd.xml: %w", err)
 	}
 
-	if primaryURL, err = GetPrimaryURL(repomd, url); err != nil {
-		return []Package{}, fmt.Errorf("Error getting primary URL: %v", err)
+	r.Repomd = result
+	return resp.StatusCode, err
+}
+
+// FetchPackages populates r.Packages with metadata of each package in repository. Returns response code and error.
+func (r *Repository) FetchPackages() (int, error) {
+	var err error
+	var primaryURL string
+	var resp *http.Response
+	var packages []Package
+
+	if _, err = r.FetchRepomd(); err != nil {
+		return 0, fmt.Errorf("error parsing repomd.xml: %w", err)
+	}
+
+	if primaryURL, err = r.getPrimaryURL(); err != nil {
+		return 0, fmt.Errorf("error getting primary URL: %w", err)
+	}
+
+	if resp, err = r.Client.Get(primaryURL); err != nil {
+		return resp.StatusCode, fmt.Errorf("GET error: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return resp.StatusCode, fmt.Errorf("status error: %w", resp.StatusCode)
+	}
+
+	if packages, err = parseCompressedXMLData(resp.Body); err != nil {
+		return resp.StatusCode, err
+	}
+	r.Packages = packages
+
+	return resp.StatusCode, nil
+}
+
+func (r *Repository) FetchGPGKey() error {
+	resp, err := fetchWithTimeout(r.URL, r.RequestTimeout)
+	if err != nil {
+		return err
+	}
+	code := resp.StatusCode
+	gpgKeyString, err := responseBodyToString(resp.Body)
+	if err != nil {
+		return err
+	}
+	if err == nil && code < 200 || code > 299 {
+		return fmt.Errorf("received http %d", code)
 	}
+	if _, openpgpErr := openpgp.ReadArmoredKeyRing(strings.NewReader(*gpgKeyString)); err != nil {
+		return openpgpErr //Bad key
+	}
+	r.GPGKey = gpgKeyString
+	return nil
+}
```

Comment:
Since GPG Keys aren't really tied to a repository, i could see this being more of a utility function, maybe in a 2nd file.

Reply:
for some background, a repo could have its url at https://packages.microsoft.com/yumrepos/azure-cli/   while its gpg key is at  https://packages.microsoft.com/keys/microsoft.asc 

these two urls aren't related